snapcraft/internal/lifecycle/_runner.py

```diff
@@ -152,83 +159,171 @@ def __init__(self, config, project_options):
         self.parts_config = config.parts
-        self._steps_run = self._init_run_states()
-
-    def _init_run_states(self):
-        steps_run = {}
+        self.steps_were_run = False
+        self._steps_run = collections.defaultdict(set)
+        self._dirty_reports = collections.defaultdict(dict)
+        summary = []
 
+        # Initialize steps run and dirty reports, so we only need to fetch
+        # them once
         for part in self.config.all_parts:
-            steps_run[part.name] = set()
-            with config.CLIConfig() as cli_config:
-                for step in common.COMMAND_ORDER:
-                    dirty_report = part.get_dirty_report(step)
-                    if dirty_report:
-                        self._handle_dirty(
-                            part, step, dirty_report, cli_config)
-                    elif not (part.should_step_run(step)):
-                        steps_run[part.name].add(step)
-                        part.notify_part_progress('Skipping {}'.format(step),
-                                                  '(already ran)')
+            part_summary = {'part': part.name}
+            for step in common.COMMAND_ORDER:
+                part_summary[step] = None
+                self._dirty_reports[part.name][step] = part.get_dirty_report(
+                    step)
+                if not part.should_step_run(step):
+                    self._steps_run[part.name].add(step)
+                    part_summary[step] = 'complete'
+                if self._dirty_reports[part.name][step]:
+                    part_summary[step] = 'dirty'
+            summary.append(part_summary)
 
-        return steps_run
+        # Print summary in debug mode
+        logger.debug(tabulate(summary, headers="keys"))
 
     def run(self, step, part_names=None):
         if part_names:
             self.parts_config.validate(part_names)
             # self.config.all_parts is already ordered, let's not lose that
             # and keep using a list.
             parts = [p for p in self.config.all_parts if p.name in part_names]
+            processed_part_names = part_names
         else:
             parts = self.config.all_parts
-            part_names = self.config.part_names
+            processed_part_names = self.config.part_names
 
         step_index = common.COMMAND_ORDER.index(step) + 1
 
-        for step in common.COMMAND_ORDER[0:step_index]:
-            if step == 'stage':
-                # XXX check only for collisions on the parts that have already
-                # been built --elopio - 20170713
-                pluginhandler.check_for_collisions(self.config.all_parts)
-            for part in parts:
-                if step not in self._steps_run[part.name]:
-                    self._run_step(step, part, part_names)
-                    self._steps_run[part.name].add(step)
+        with config.CLIConfig() as cli_config:
+            for current_step in common.COMMAND_ORDER[0:step_index]:
```

Comment:
This CLIConfig context feels to broad like this, but I cannot think of an easier way using a context which would not require opening and closing all the time. Is loading outside of the context a better alternative here, oh right, the context closes it, bah.
This would look nicer if we had `defer` or maybe if it decorated `run` somehow ;-)

Reply:
Wow, you just echoed the exact thought process I had when writing this :laughing: .